### PR TITLE
[FLINK-12760] [runtime] Implement ExecutionGraph to InputsLocationsRetriever Adapter

### PR DESCRIPTION
**Pull request by shuai-xu**
## What is the purpose of the change

*This pull request is based on #8486 , it implements a InputsLocationsRetriever based on ExecutionGraph.*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added unit tests in EGBasedInputsLocationsRetrieverTest.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetriever.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+		} else {
+			return Optional.empty();
+		}
+	}
+
+	private ExecutionVertex getExecutionVertex(ExecutionVertexID executionVertexId) {
+
+		ExecutionJobVertex ejv = executionGraph.getJobVertex(executionVertexId.getJobVertexId());
+		if (ejv == null || ejv.getParallelism() <= executionVertexId.getSubtaskIndex()) {
+			throw new IllegalArgumentException(String.format("Failed to find execution {} in execution graph.", executionVertexId));
```
Comment (GJL) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetriever.java]: `{}` as placeholders don't work here

**Comment by GJL** [comment]
Can you rebase?

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetriever.java (hunk 1 above)
Comment (shuai-xu) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetriever.java]: Done

**Comment by shuai-xu** [comment]
@GJL Of course, I have rebased it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetrieverTest.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+		ExecutionVertexID executionVertexId = new ExecutionVertexID(jobVertex.getID(), 0);
+		Optional<CompletableFuture<TaskManagerLocation>> taskManagerLocation =
+				inputsLocationsRetriever.getTaskManagerLocation(executionVertexId);
+
+		assertFalse(taskManagerLocation.isPresent());
+	}
+
+	/**
+	 * Tests that when exection is not scheduled, getting task manager location will return null.
```
Comment (eaglewatcherwb) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetrieverTest.java]: typo: execution
The doc seems mismatch with the unit test.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetrieverTest.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+		Collection<Collection<ExecutionVertexID>> producersOfProducer1 =
+				inputsLocationsRetriever.getConsumedResultPartitionsProducers(evIdOfProducer1);
+		Collection<Collection<ExecutionVertexID>> producersOfProducer2 =
+				inputsLocationsRetriever.getConsumedResultPartitionsProducers(evIdOfProducer2);
+		Collection<Collection<ExecutionVertexID>> producersOfConsumer =
+				inputsLocationsRetriever.getConsumedResultPartitionsProducers(evIdOfConsumer);
+
+		assertThat(producersOfProducer1, is(empty()));
+		assertThat(producersOfProducer2, is(empty()));
+		assertThat(producersOfConsumer, contains(Arrays.asList(evIdOfProducer1), Arrays.asList(evIdOfProducer2)));
```
Comment (eaglewatcherwb) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetrieverTest.java]: `Arrays.asList` with only one argument could be replaced with `Collections.singletonList`

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetrieverTest.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+		assertFalse(taskManagerLocation.isPresent());
+	}
+
+	/**
+	 * Tests that it can get the task manager location in an Execution.
+	 */
+	@Test
+	public void testGetTaskManagerLocationWhenScheduled() throws Exception {
+		final JobVertex jobVertex = ExecutionGraphTestUtils.createNoOpVertex(1);
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetrieverTest.java]: The method `scheduleForExecution()` should be already considered legacy.

Imo, we should do something like:

```
		final JobVertex jobVertex = ExecutionGraphTestUtils.createNoOpVertex(1);

		final TestingLogicalSlot testingLogicalSlot = new TestingLogicalSlot();
		final ExecutionGraph eg = ExecutionGraphTestUtils.createSimpleTestGraph(new JobID(), jobVertex);
		final EGBasedInputsLocationsRetriever inputsLocationsRetriever = new EGBasedInputsLocationsRetriever(eg);

		final ExecutionVertex onlyExecutionVertex = eg.getAllExecutionVertices().iterator().next();
		onlyExecutionVertex.deployToSlot(testingLogicalSlot);

		ExecutionVertexID consumerVertexId = new ExecutionVertexID(jobVertex.getID(), 0);
		Optional<CompletableFuture<TaskManagerLocation>> taskManagerLocationOptional = inputsLocationsRetriever.getTaskManagerLocation(consumerVertexId);

		assertTrue(taskManagerLocationOptional.isPresent());

		final CompletableFuture<TaskManagerLocation> taskManagerLocationFuture = taskManagerLocationOptional.get();
		assertEquals(testingLogicalSlot.getTaskManagerLocation(), taskManagerLocationFuture.get());
```

What do you think?

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetrieverTest.java (hunk 3 above)
Comment (GJL) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetrieverTest.java]: This line causes a warning
image
Also see: http://www.tothenew.com/blog/why-is-generic-array-creation-not-allowed-in-java/

Hence, I would propose:
```
		assertThat(producersOfConsumer, hasSize(2));
		assertThat(producersOfConsumer, hasItem(Collections.singletonList(evIdOfProducer1)));
		assertThat(producersOfConsumer, hasItem(Collections.singletonList(evIdOfProducer2)));
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetriever.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * An implementation of {@link InputsLocationsRetriever} based on the {@link ExecutionGraph}.
+ */
+public class EGBasedInputsLocationsRetriever implements InputsLocationsRetriever {
```
Comment (GJL) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetriever.java]: I would spell it out. Abbreviations can be hard to understand.

`ExecutionGraphToInputsLocationsRetrieverAdapter`

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetriever.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+			return Optional.of(ev.getCurrentTaskManagerLocationFuture());
+		} else {
+			return Optional.empty();
+		}
+	}
+
+	private ExecutionVertex getExecutionVertex(ExecutionVertexID executionVertexId) {
+
+		ExecutionJobVertex ejv = executionGraph.getJobVertex(executionVertexId.getJobVertexId());
+		if (ejv == null || ejv.getParallelism() <= executionVertexId.getSubtaskIndex()) {
```
Comment (GJL) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetriever.java]: Consider 
```
checkState(ejv != null && ejv.getParallelism() > executionVertexId.getSubtaskIndex(),
			"Failed to find execution %s in execution graph.", executionVertexId);
```
which is shorter

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetriever.java (hunk 5 above)
Comment (shuai-xu) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetriever.java]: Done

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetriever.java (hunk 6 above)
Comment (shuai-xu) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetriever.java]: OK

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetrieverTest.java (hunk 4 above)
Comment (shuai-xu) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetrieverTest.java]: OK， reasonable

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetrieverTest.java (hunk 3 above)
Comment (shuai-xu) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetrieverTest.java]: OK

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetriever.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+			throw new IllegalArgumentException(String.format("Failed to find execution %s in execution graph.", executionVertexId));
```
[the remark this answers is not in the document]
Comment (shuai-xu) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetriever.java]: Done

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/ExecutionGraphToInputsLocationsRetrieverAdapterTest.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+		assertThat(producersOfConsumer, hasSize(2));
+		assertThat(producersOfConsumer, hasItem(Collections.singletonList(evIdOfProducer1)));
+		assertThat(producersOfConsumer, hasItem(Collections.singletonList(evIdOfProducer2)));
+	}
+
+	/**
+	 * Tests that when execution is not scheduled, getting task manager location will return null.
+	 */
+	@Test
+	public void testGetNullTaskManagerLocationIfNotScheduled() throws Exception {
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/ExecutionGraphToInputsLocationsRetrieverAdapterTest.java]: Maybe `[...]EmptyTaskManagerLocationIfVertexNotScheduled()'

Comment (shuai-xu) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/ExecutionGraphToInputsLocationsRetrieverAdapterTest.java]: Done

**Comment by shuai-xu** [comment]
@GJL ， It seems the failure of tests is due to the travis system, not by the code.

**Comment by GJL** [comment]
Merging.

**Pull request closed without merging**